File: rust/crates/test/src/de.rs
use crate::schema;

#[test]
fn deserialize_empty() {
  let m = schema::Test::deserialize(vec![0]).unwrap();
  assert_eq!(m, schema::Empty {}.into());
}

#[test]
fn deserialize_u8() {
  let m = schema::Test::deserialize(vec![1, 69]).unwrap();
  assert_eq!(m, schema::U8 { value: 69 }.into());
}

#[test]
fn deserialize_bool() {
  let m = schema::Test::deserialize(vec![12, 1]).unwrap();
  assert_eq!(m, schema::Bool { value: true }.into());
}

#[test]
fn deserialize_u16() {
  let m = schema::Test::deserialize(vec![2, 0x34, 0x12]).unwrap();
  assert_eq!(m, schema::U16 { value: 0x1234 }.into());
}

#[test]
fn deserialize_u32() {
  let m = schema::Test::deserialize(vec![3, 0x78, 0x56, 0x34, 0x12]).unwrap();
  assert_eq!(m, schema::U32 { value: 0x12345678 }.into());
}

#[test]
fn deserialize_u64() {
  let m =
    schema::Test::deserialize(vec![4, 0xf0, 0xde, 0xbc, 0x9a, 0x78, 0x56, 0x34, 0x12]).unwrap();
  assert_eq!(
    m,
    schema::U64 {
      value: 0x123456789ABCDEF0,
    }
    .into()
  );
}

#[test]
fn deserialize_i8() {
  let m = schema::Test::deserialize(vec![5, 69]).unwrap();
  assert_eq!(m, schema::I8 { value: 69 }.into());
}

#[test]
fn deserialize_i16() {
  let m = schema::Test::deserialize(vec![6, 0x34, 0x12]).unwrap();
  assert_eq!(m, schema::I16 { value: 0x1234 }.into());
}

#[test]
fn deserialize_i32() {
  let m = schema::Test::deserialize(vec![7, 0x78, 0x56, 0x34, 0x12]).unwrap();
  assert_eq!(m, schema::I32 { value: 0x12345678 }.into());
}

#[test]
fn deserialize_i64() {
  let m =
    schema::Test::deserialize(vec![8, 0xf0, 0xde, 0xbc, 0x9a, 0x78, 0x56, 0x34, 0x12]).unwrap();
  assert_eq!(
    m,
    schema::I64 {
      value: 0x123456789ABCDEF0,
    }
    .into()
  );
}

#[test]
fn deserialize_i8_negative() {
  let m = schema::Test::deserialize(vec![5, 187]).unwrap();
  assert_eq!(m, schema::I8 { value: -69 }.into());
}

#[test]
fn deserialize_i16_negative() {
  let m = schema::Test::deserialize(vec![6, 0xCC, 0xED]).unwrap();
  assert_eq!(m, schema::I16 { value: -0x1234 }.into());
}

#[test]
fn deserialize_i32_negative() {
  let m = schema::Test::deserialize(vec![7, 0x88, 0xA9, 0xCB, 0xED]).unwrap();
  assert_eq!(m, schema::I32 { value: -0x12345678 }.into());
}

#[test]
fn deserialize_i64_negative() {
  let m =
    schema::Test::deserialize(vec![8, 0x10, 0x21, 0x43, 0x65, 0x87, 0xA9, 0xCB, 0xED]).unwrap();
  assert_eq!(
    m,
    schema::I64 {
      value: -0x123456789ABCDEF0,
    }
    .into()
  );
}

#[test]
fn deserialize_string() {
  let m = schema::Test::deserialize(vec![9, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x21, 0x0]).unwrap();
  assert_eq!(
    m,
    schema::Str {
      value: "hello!".into(),
    }
    .into()
  );
}

#[test]
fn deserialize_f32() {
  let m = schema::Test::deserialize(vec![10, 0xdb, 0x0f, 0x49, 0x40]).unwrap();
  assert_eq!(
    m,
    schema::F32 {
      value: std::f32::consts::PI,
    }
    .into()
  );
}

#[test]
fn deserialize_f64() {
  let m =
    schema::Test::deserialize(vec![11, 0x18, 0x2D, 0x44, 0x54, 0xFB, 0x21, 0x09, 0x40]).unwrap();
  assert_eq!(
    m,
    schema::F64 {
      value: std::f64::consts::PI,
    }
    .into()
  );
}

#[test]
fn no_zero_termination() {
  let m = schema::Test::deserialize(vec![9, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x21]);
  assert!(m.is_none());
}

File: lua/tests/de.lua
--- @diagnostic disable: need-check-nil

local Test = require("tests.gen.test")

return {
	testDeserializeEmpty = function()
		local m = Test.deserialize(string.char(0))
		Lu.assertNotIsNil(m)
	end,

	testDeserializeU8 = function()
		local m = Test.deserialize(string.char(1, 69))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.U8.new(69):_wrap():_serialize())
	end,

	testDeserializeBool = function()
		local m = Test.deserialize(string.char(12, 1))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.Bool.new(true):_wrap():_serialize())
	end,

	testDeserializeU16 = function()
		local m = Test.deserialize(string.char(2, 0x34, 0x12))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.U16.new(0x1234):_wrap():_serialize())
	end,

	testDeserializeU32 = function()
		local m = Test.deserialize(string.char(3, 0x78, 0x56, 0x34, 0x12))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.U32.new(0x12345678):_wrap():_serialize())
	end,

	testDeserializeI8 = function()
		local m = Test.deserialize(string.char(5, 69))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I8.new(69):_wrap():_serialize())
	end,

	testDeserializeI16 = function()
		local m = Test.deserialize(string.char(6, 0x34, 0x12))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I16.new(0x1234):_wrap():_serialize())
	end,

	testDeserializeI32 = function()
		local m = Test.deserialize(string.char(7, 0x78, 0x56, 0x34, 0x12))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I32.new(0x12345678):_wrap():_serialize())
	end,

	testDeserializeI8Negative = function()
		local m = Test.deserialize(string.char(5, 187))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I8.new(-69):_wrap():_serialize())
	end,

	testDeserializeI16Negative = function()
		local m = Test.deserialize(string.char(6, 0xcc, 0xed))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I16.new(-0x1234):_wrap():_serialize())
	end,

	testDeserializeI32Negative = function()
		local m = Test.deserialize(string.char(7, 0x88, 0xa9, 0xcb, 0xed))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.I32.new(-0x12345678):_wrap():_serialize())
	end,

	testDeserializeString = function()
		local m = Test.deserialize(string.char(9, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x21, 0))
		Lu.assertNotIsNil(m)
		Lu.assertEquals(m:_serialize(), Test.Str.new("hello!"):_wrap():_serialize())
	end,

	testDeserializeF32 = function()
		local m = Test.deserialize(string.char(10, 0xdb, 0x0f, 0x49, 0x40))
		Lu.assertNotIsNil(m)
	end,

	testDeserializeF64 = function()
		local m = Test.deserialize(string.char(11, 0x18, 0x2d, 0x44, 0x54, 0xfb, 0x21, 0x09, 0x40))
		Lu.assertNotIsNil(m)
	end,

	testNoZeroTermination = function()
		local m = Test.deserialize(string.char(9, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x21))
		Lu.assertIsNil(m)
	end,

	testLength = function()
		local m, l = Test.deserialize(
			string.char(130, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
		)
		Lu.assertEquals(l, 16)
	end,
}
